datasync/chaindatafetcher/chaindata_fetcher.go

```diff
@@ -107,9 +127,102 @@ func (f *ChainDataFetcher) Stop() error {
 	logger.Info("wait for all goroutines to be terminated...")
 	f.wg.Wait()
 	logger.Info("terminated all goroutines for chaindatafetcher")
+	f.started = false
+	return nil
+}
+
+func (f *ChainDataFetcher) start() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.started {
+		return errors.New("the chaindata fetcher is already started")
+	}
+
+	f.chainSub = f.blockchain.SubscribeChainEvent(f.chainCh)
+	// TODO-ChainDataFetcher add logic to request from checkpoint to the first posted event after start.
+
+	f.started = true
 	return nil
 }
 
+func (f *ChainDataFetcher) stop() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if !f.started {
+		return errors.New("the chaindata fetcher is already stopped")
+	}
+
+	f.chainSub.Unsubscribe()
+	f.started = false
+	return nil
+}
+
+func (f *ChainDataFetcher) startRange(start, end int64, reqType requestType) error {
+	f.rangeMu.Lock()
+	defer f.rangeMu.Unlock()
+	if f.rangeStarted {
+		return errors.New("the chaindata fetcher is already started with range")
+	}
+	f.rangeStarted = true
+	defer func() { f.rangeStarted = false }()
+
+	events, err := f.makeChainEvents(start, end)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		f.reqCh <- &request{
+			reqType: requestTypeTransaction,
+			event:   e,
+		}
+		// TODO-ChainDataFetcher add stop logic while processing the events.
+	}
+	return nil
+}
+
+func (f *ChainDataFetcher) stopRange() error {
+	// TODO-ChainDataFetcher add logic for stopping
+	return nil
+}
+
+func (f *ChainDataFetcher) makeChainEvents(start, end int64) ([]blockchain.ChainEvent, error) {
```

Comment:
In next time, please change the code that each handler makes the event itself.

Reply:
I added comment. I will do this later on. 0bad014154444e6d3381154d41c8c131a1b9a085